File: ios/Shared/EdgeSearch/EdgeModels.swift
import Foundation

/// One indexed maintenance-doc section from the edge index.
struct EdgeChunk: Identifiable, Sendable {
    let id: String            // ChunkId
    let documentId: String
    let sectionTitle: String
    let contentText: String
    let errorCodes: [String]
    let toolsRequired: [String]

    /// Decodes the `Metadata` JSON column: { "error_codes":[…], "tools_required":[…] }.
    struct Metadata {
        let errorCodes: [String]
        let toolsRequired: [String]
        init(json: String) {
            struct Raw: Decodable { let error_codes: [String]?; let tools_required: [String]? }
            let raw = try? JSONDecoder().decode(Raw.self, from: Data(json.utf8))
            errorCodes = raw?.error_codes ?? []
            toolsRequired = raw?.tools_required ?? []
        }
    }
}

/// A ranked hit: the chunk and its cosine score (0…1).
struct EdgeSearchHit: Identifiable, Sendable {
    let chunk: EdgeChunk
    let score: Float
    var id: String { chunk.id }
}

/// Errors surfaced by the edge engine; rendered as graceful "unavailable"/inline cards.
enum EdgeSearchError: LocalizedError {
    case assetMissing(String)
    case dbOpenFailed
    case inferenceFailed

    var errorDescription: String? {
        switch self {
        case .assetMissing(let f): return "Missing bundled asset: \(f)."
        case .dbOpenFailed: return "Couldn't open the compiled search index."
        case .inferenceFailed: return "On-device embedding failed (Core ML)."
        }
    }
}

File: android/app/src/main/kotlin/io/dotnetnativeinterop/model/EdgeModels.kt
package io.dotnetnativeinterop.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/** Shared lenient JSON for EVS payloads (index Metadata + fixtures). */
internal val evsJson: Json = Json { ignoreUnknownKeys = true }

/** The Metadata JSON column in edge-index.db: {"error_codes":[…],"tools_required":[…]}. */
@Serializable
public data class EdgeMetadata(
    @SerialName("error_codes") val errorCodes: List<String> = emptyList(),
    @SerialName("tools_required") val toolsRequired: List<String> = emptyList(),
)

/** One indexed chunk from edge-index.db (embedding is the L2-normalized 384-d vector). */
public data class EdgeChunk(
    val chunkId: String,
    val documentId: String,
    val sectionTitle: String,
    val contentText: String,
    val errorCodes: List<String>,
    val toolsRequired: List<String>,
    val embedding: FloatArray,
)

/** A ranked search hit. */
public data class EdgeHit(val chunk: EdgeChunk, val score: Float)
